interpreter/cling/lib/Utils/ASTImportSource.cpp

```diff
@@ -0,0 +1,189 @@
+          }
+        } else {
+          Decl *declToImport = llvm::cast<Decl>(*lookup_result.data());
+          ImportDecl(declToImport, importer, childDeclName,
+                     parentDeclName, childCurrentDeclContext);
+        }
+      }
+      return true;
+    }
+
+    ///\brief This is the most important function of the class ASTImportSource
+    /// since from here initiates the lookup and import part of the missing
+    /// Decl(s) (Contexts).
+    ///
+    bool ASTImportSource::FindExternalVisibleDeclsByName(
+      const DeclContext *childCurrentDeclContext, DeclarationName childDeclName) {
+
+      assert(childCurrentDeclContext->hasExternalVisibleStorage() &&
+             "DeclContext has no visible decls in storage");
+
+      /* clang will call FindExternalVisibleDeclsByName with an
+       * IdentifierInfo valid for the child interpreter. Get the
+       * IdentifierInfo's StringRef representation.
+       * Get the identifier info from the parent interpreter
+       * for this Name. */
+      llvm::StringRef name(childDeclName.getAsString());
+      IdentifierTable &parentIdentifierTable =
+        m_parent_Interp->getCI()->getASTContext().Idents;
+      IdentifierInfo &parentIdentifierInfo = parentIdentifierTable.get(name);
+      DeclarationName parentDeclName(&parentIdentifierInfo);
+
+      /* Check if we have already imported this Decl (Context). */
+      if (m_DeclName_map.find(childDeclName.getAsString()) != m_DeclName_map.end())
+        return true;
+
+      DeclContext::lookup_result lookup_result;
+      /* If we are not looking for this Name in the Translation Unit
+       * but instead inside a namespace, */
+      if (!childCurrentDeclContext->isTranslationUnit()) {
+        /* Search in the map of the stored Decl Contexts for this
+         * namespace. */
+        if (m_DeclContexts_map.find(childCurrentDeclContext) != m_DeclContexts_map.end()) {
+          /* If childCurrentDeclContext was found before and is already in the map,
+           * then do the lookup using the stored pointer. */
+          DeclContext *originalDeclContext =
+            m_DeclContexts_map.find(childCurrentDeclContext)->second;
+
+          Decl *fromDeclContext = Decl::castFromDeclContext(originalDeclContext);
+          ASTContext &from_ASTContext = fromDeclContext->getASTContext();
+
+          Decl *toDeclContext = Decl::castFromDeclContext(childCurrentDeclContext);
+          ASTContext &to_ASTContext = toDeclContext->getASTContext();
+
+          lookup_result = originalDeclContext->lookup(parentDeclName);
+          /* Do the import */
+          if (Import(lookup_result, from_ASTContext, to_ASTContext,
+                     childCurrentDeclContext, childDeclName, parentDeclName))
+            return true;
+        }
+      } else { /* Otherwise search in the Translation Unit. */
+        ASTContext &from_ASTContext = m_parent_Interp->getCI()->getASTContext();
```

Comment:
I assume that this will be collapsed with the other `if` case above - not reviewed.